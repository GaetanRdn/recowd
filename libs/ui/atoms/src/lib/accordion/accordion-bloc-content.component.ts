import { Component, HostBinding } from '@angular/core';
import { collapsable } from '@recowd/ui-animations';

@Component({
  selector: 'rc-accordion-bloc-content',
  standalone: true,
  template: ` <span class="rc-accordion-bloc-content-wrapper"
    ><ng-content></ng-content
  ></span>`,
  animations: [collapsable],
})
export class AccordionBlocContentComponent {
  @HostBinding('class.rc-collapsed') public collapsed = true;

  @HostBinding('@collapsable') get collapsable(): 'collapsed' | 'expanded' {
    return this.collapsed ? 'collapsed' : 'expanded';
  }
}
