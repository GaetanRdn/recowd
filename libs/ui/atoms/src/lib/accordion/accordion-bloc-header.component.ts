import { Component, EventEmitter, HostListener, Output } from '@angular/core';
import { MaterialIconDirective } from '../material-icon/material-icon.directive';
import { AutoUnsubscribe } from '@recowd/utility-types';
import { rotation180Animation } from '@recowd/ui-animations';

@Component({
  selector: 'rc-accordion-bloc-header',
  host: {
    class: 'rc-accordion-bloc-header', // TODO remove
  },
  standalone: true,
  imports: [MaterialIconDirective],
  template: ` <ng-content></ng-content>
    <rc-material-icon [@rotate]="selected ? 'rotated' : 'default'">
      expand_more
    </rc-material-icon>`,
  animations: [rotation180Animation],
})
@AutoUnsubscribe()
export class AccordionBlocHeaderComponent {
  public selected = false;

  @Output() public readonly selectChange: EventEmitter<void> =
    new EventEmitter<void>();

  @HostListener('click')
  private click(): void {
    this.selectChange.emit();
    this.selected = !this.selected;
  }
}
