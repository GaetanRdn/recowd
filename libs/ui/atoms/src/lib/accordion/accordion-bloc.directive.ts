import {
  AfterContentInit,
  ChangeDetectorRef,
  ContentChild,
  Directive,
  inject,
  Input,
  OnChanges,
} from '@angular/core';
import { AccordionBlocContentComponent } from './accordion-bloc-content.component';
import { AccordionBlocHeaderComponent } from './accordion-bloc-header.component';
import { Subscription, tap } from 'rxjs';
import {
  AutoUnsubscribe,
  BooleanInput,
  CoerceBoolean,
  NgChanges,
} from '@recowd/utility-types';

@Directive({
  selector: '[rcAccordionBloc], rc-accordion-bloc',
  host: {
    class: 'rc-accordion-bloc',
  },
  standalone: true,
})
@AutoUnsubscribe()
export class AccordionBlocDirective implements AfterContentInit, OnChanges {
  static ngAcceptInputType_collapsed: BooleanInput;

  @Input() @CoerceBoolean() public collapsed = true;

  @ContentChild(AccordionBlocHeaderComponent)
  public readonly header!: AccordionBlocHeaderComponent;

  @ContentChild(AccordionBlocContentComponent)
  public readonly content!: AccordionBlocContentComponent;

  private readonly _cdr: ChangeDetectorRef = inject(ChangeDetectorRef);

  private readonly _subscriptions: Subscription[] = [];

  public ngOnChanges(changes: NgChanges<AccordionBlocDirective>): void {
    if (changes.collapsed && !changes.collapsed.isFirstChange()) {
      this.content.collapsed = this.collapsed;
    }
  }

  public ngAfterContentInit(): void {
    this._subscriptions[this._subscriptions.length] = this.header.selectChange
      .pipe(tap(() => (this.content.collapsed = !this.content.collapsed)))
      .subscribe(() => this._cdr.markForCheck());

    if (this.collapsed !== this.content.collapsed) {
      this.content.collapsed = this.collapsed;
    }
  }
}
