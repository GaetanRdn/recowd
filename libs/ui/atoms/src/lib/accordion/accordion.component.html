<ng-content select="[rcAccordionBloc], rc-accordion-bloc"></ng-content>
