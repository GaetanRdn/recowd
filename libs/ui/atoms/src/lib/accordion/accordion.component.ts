import { ChangeDetectionStrategy, Component } from '@angular/core';

@Component({
  selector: 'rc-accordion',
  templateUrl: './accordion.component.html',
  styleUrls: ['./accordion.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class AccordionComponent {}
