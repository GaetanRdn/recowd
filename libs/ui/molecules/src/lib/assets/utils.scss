@function map-get-deep($map, $level, $subLevel) {
  @return map-get($map, $level, $subLevel);
}

@mixin use-typography($typography) {
  @each $key in font-size, font-family, font-weight, line-height {
    #{$key}: map-get($typography, $key);
  }
}
