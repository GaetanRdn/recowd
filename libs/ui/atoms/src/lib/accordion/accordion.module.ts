import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AccordionComponent } from './accordion.component';
import { AccordionBlocDirective } from './accordion-bloc.directive';
import { AccordionBlocHeaderComponent } from './accordion-bloc-header.component';
import { AccordionBlocContentComponent } from './accordion-bloc-content.component';

@NgModule({
  declarations: [AccordionComponent],
  exports: [
    AccordionComponent,
    AccordionBlocDirective,
    AccordionBlocHeaderComponent,
    AccordionBlocContentComponent,
  ],
  imports: [
    CommonModule,
    AccordionBlocDirective,
    AccordionBlocHeaderComponent,
    AccordionBlocContentComponent,
  ],
})
export class AccordionModule {}
