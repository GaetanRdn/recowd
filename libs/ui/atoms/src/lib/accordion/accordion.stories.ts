import { Args, Meta, moduleMetadata, Story } from '@storybook/angular';
import { AccordionComponent } from './accordion.component';
import { AccordionModule } from './accordion.module';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';

export default {
  title: 'atoms/accordion',
  parameters: {
    jest: 'accordion',
  },
  decorators: [
    moduleMetadata({
      imports: [AccordionModule, BrowserAnimationsModule],
    }),
  ],
} as Meta<AccordionComponent>;

const templateSingleBloc: Story<AccordionComponent> = (args: Args) => ({
  props: {
    collapsed: args['collapsed'],
  },
  template: `<rc-accordion>
    <rc-accordion-bloc [collapsed]="collapsed">
      <rc-accordion-bloc-header>Example</rc-accordion-bloc-header>
      <rc-accordion-bloc-content>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam vulputate mauris a mattis euismod. Suspendisse ultrices fringilla ipsum. Suspendisse sit amet varius lectus. Mauris erat elit, euismod a laoreet sit amet, finibus vestibulum augue. Cras non consectetur orci. Praesent elit nisi, volutpat sit amet tempus non, tincidunt a felis.</rc-accordion-bloc-content>
    </rc-accordion-bloc>
  </rc-accordion>`,
});

export const SingleBloc = templateSingleBloc.bind({});
SingleBloc.args = {
  collapsed: true,
};

const templateMultiBlocs: Story<AccordionComponent> = (args: Args) => ({
  template: `<rc-accordion>
    <rc-accordion-bloc>
      <rc-accordion-bloc-header>Example</rc-accordion-bloc-header>
      <rc-accordion-bloc-content>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam vulputate mauris a mattis euismod. Suspendisse ultrices fringilla ipsum. Suspendisse sit amet varius lectus. Mauris erat elit, euismod a laoreet sit amet, finibus vestibulum augue. Cras non consectetur orci. Praesent elit nisi, volutpat sit amet tempus non, tincidunt a felis.</rc-accordion-bloc-content>
    </rc-accordion-bloc>

    <rc-accordion-bloc>
      <rc-accordion-bloc-header>Example bis</rc-accordion-bloc-header>
      <rc-accordion-bloc-content>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam vulputate mauris a mattis euismod. Suspendisse ultrices fringilla ipsum. Suspendisse sit amet varius lectus. Mauris erat elit, euismod a laoreet sit amet, finibus vestibulum augue. Cras non consectetur orci. Praesent elit nisi, volutpat sit amet tempus non, tincidunt a felis.</rc-accordion-bloc-content>
    </rc-accordion-bloc>

    <rc-accordion-bloc>
      <rc-accordion-bloc-header>Example ter</rc-accordion-bloc-header>
      <rc-accordion-bloc-content>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam vulputate mauris a mattis euismod. Suspendisse ultrices fringilla ipsum. Suspendisse sit amet varius lectus. Mauris erat elit, euismod a laoreet sit amet, finibus vestibulum augue. Cras non consectetur orci. Praesent elit nisi, volutpat sit amet tempus non, tincidunt a felis.</rc-accordion-bloc-content>
    </rc-accordion-bloc>
  </rc-accordion>`,
});

export const MultiBlocs = templateMultiBlocs.bind({});
