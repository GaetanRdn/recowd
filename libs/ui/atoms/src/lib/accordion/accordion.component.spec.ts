import { TestBed, waitForAsync } from '@angular/core/testing';

import { AccordionComponent } from './accordion.component';
import { TemplateLookup } from '@recowd/test/utils';
import { Component } from '@angular/core';
import { AccordionModule } from './accordion.module';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';

describe('AccordionComponent', () => {
  let templateLookup: TemplateLookup<SingleBlocHostComponent>;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [SingleBlocHostComponent],
      imports: [AccordionModule, NoopAnimationsModule],
    }).compileComponents();
  }));

  beforeEach(
    () => (templateLookup = new TemplateLookup(SingleBlocHostComponent))
  );

  it('should create default closed accordion bloc', waitForAsync(() => {
    templateLookup.detectChanges();
    expect(templateLookup.firstChildElement).toMatchSnapshot();
  }));

  it('should expand on header click', waitForAsync(() => {
    templateLookup.detectChanges();

    templateLookup.query('rc-accordion-bloc-header').click();
    templateLookup.detectChanges();

    expect(templateLookup.firstChildElement).toMatchSnapshot();
  }));

  it('should be opened', waitForAsync(() => {
    templateLookup.hostComponent.collapsed = false;

    templateLookup.detectChanges();

    expect(templateLookup.firstChildElement).toMatchSnapshot();
  }));

  it('should be opened on collapsed change', waitForAsync(() => {
    templateLookup.hostComponent.collapsed = true;
    templateLookup.detectChanges();

    templateLookup.hostComponent.collapsed = false;
    templateLookup.detectChanges();

    expect(templateLookup.firstChildElement).toMatchSnapshot();
  }));
});

@Component({
  template: ` <rc-accordion>
    <rc-accordion-bloc [collapsed]="collapsed">
      <rc-accordion-bloc-header>Title</rc-accordion-bloc-header>
      <rc-accordion-bloc-content>Content</rc-accordion-bloc-content>
    </rc-accordion-bloc>
  </rc-accordion>`,
})
class SingleBlocHostComponent {
  public collapsed = false;
}

@Component({
  template: ` <rc-accordion>
    <rc-accordion-bloc>
      <rc-accordion-bloc-header>Title</rc-accordion-bloc-header>
      <rc-accordion-bloc-content>Content</rc-accordion-bloc-content>
    </rc-accordion-bloc>
    <rc-accordion-bloc>
      <rc-accordion-bloc-header>Title2</rc-accordion-bloc-header>
      <rc-accordion-bloc-content>Content2</rc-accordion-bloc-content>
    </rc-accordion-bloc>
    <rc-accordion-bloc>
      <rc-accordion-bloc-header>Title3</rc-accordion-bloc-header>
      <rc-accordion-bloc-content>Content3</rc-accordion-bloc-content>
    </rc-accordion-bloc>
  </rc-accordion>`,
})
class MultiBlocsHostComponent {}
