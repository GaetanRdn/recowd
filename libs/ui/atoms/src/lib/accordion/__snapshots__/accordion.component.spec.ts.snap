// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionComponent should be opened 1`] = `
<rc-accordion>
  <rc-accordion-bloc
    class="rc-accordion-bloc"
  >
    <rc-accordion-bloc-header
      class="rc-accordion-bloc-header"
    >
      Title
      <rc-material-icon
        class="ng-trigger ng-trigger-rotate material-icons"
        style=""
      >
         expand_more 
      </rc-material-icon>
    </rc-accordion-bloc-header>
    <rc-accordion-bloc-content
      class="ng-trigger ng-trigger-collapsable"
      style="opacity:1;0:opacity;"
    >
      <span
        class="rc-accordion-bloc-content-wrapper"
      >
        Content
      </span>
    </rc-accordion-bloc-content>
  </rc-accordion-bloc>
</rc-accordion>
`;

exports[`AccordionComponent should be opened on collapsed change 1`] = `
<rc-accordion>
  <rc-accordion-bloc
    class="rc-accordion-bloc"
  >
    <rc-accordion-bloc-header
      class="rc-accordion-bloc-header"
    >
      Title
      <rc-material-icon
        class="ng-trigger ng-trigger-rotate material-icons"
        style=""
      >
         expand_more 
      </rc-material-icon>
    </rc-accordion-bloc-header>
    <rc-accordion-bloc-content
      class="ng-trigger ng-trigger-collapsable ng-animating"
      style=""
    >
      <span
        class="rc-accordion-bloc-content-wrapper"
      >
        Content
      </span>
    </rc-accordion-bloc-content>
  </rc-accordion-bloc>
</rc-accordion>
`;

exports[`AccordionComponent should create default closed accordion bloc 1`] = `
<rc-accordion>
  <rc-accordion-bloc
    class="rc-accordion-bloc"
  >
    <rc-accordion-bloc-header
      class="rc-accordion-bloc-header"
    >
      Title
      <rc-material-icon
        class="ng-trigger ng-trigger-rotate material-icons"
        style=""
      >
         expand_more 
      </rc-material-icon>
    </rc-accordion-bloc-header>
    <rc-accordion-bloc-content
      class="ng-trigger ng-trigger-collapsable"
      style="opacity:1;0:opacity;"
    >
      <span
        class="rc-accordion-bloc-content-wrapper"
      >
        Content
      </span>
    </rc-accordion-bloc-content>
  </rc-accordion-bloc>
</rc-accordion>
`;

exports[`AccordionComponent should expand on header click 1`] = `
<rc-accordion>
  <rc-accordion-bloc
    class="rc-accordion-bloc"
  >
    <rc-accordion-bloc-header
      class="rc-accordion-bloc-header"
    >
      Title
      <rc-material-icon
        class="ng-trigger ng-trigger-rotate material-icons ng-animating"
        style=""
      >
         expand_more 
      </rc-material-icon>
    </rc-accordion-bloc-header>
    <rc-accordion-bloc-content
      class="ng-trigger ng-trigger-collapsable rc-collapsed ng-animating"
      style=""
    >
      <span
        class="rc-accordion-bloc-content-wrapper"
      >
        Content
      </span>
    </rc-accordion-bloc-content>
  </rc-accordion-bloc>
</rc-accordion>
`;
